globus_action_provider_tools/data_types.py

```diff
@@ -84,31 +86,93 @@ class ActionProviderDescription:
     event_types: Optional[List[EventType]] = None
 
 
-@dataclass
-class ActionRequest:
-    request_id: str
-    body: Dict[str, Any]
-    label: Optional[str] = None
-    monitor_by: Optional[List[str]] = None
-    manage_by: Optional[List[str]] = None
-    allowed_clients: Optional[List[str]] = None
-    deadline: Optional[str] = None
-    release_after: Optional[str] = None
-
-
-@dataclass
-class ActionStatus:
-    status: ActionStatusValue
-    creator_id: str
-    action_id: str = field(default_factory=shortish_id)
-    start_time: str = field(default_factory=now_isoformat)
-    label: Optional[str] = None
-    monitor_by: Optional[Iterable[str]] = None
-    manage_by: Optional[Iterable[str]] = None
-    completion_time: Optional[str] = None
-    release_after: Optional[str] = None
-    display_status: Optional[str] = None
-    details: Optional[Dict[str, Any]] = None
+class ActionRequest(BaseModel):
+    request_id: str = Field(
+        ...,
+        description="A unique identifier representing the request to start an action. Multiple uses of the same request_id must have the same content or they will be rejected. Only one instance of the operation will be executed, so requests with the same request_id may be repeated to attempt to guarantee execution of an action",
+    )
+    body: Dict[str, Any] = Field(
+        ...,
+        description="The Action Provider-specific content describing the action to run. The format for the body is provided in the input_schema field of the Action Provider Description",
+    )
+    label: Optional[str] = Field(
+        None,
+        description="A short human presentable description of the Action requested",
+        min_length=1,
+        max_length=64,
+    )
+    deadline: Optional[datetime.datetime] = Field(
+        None,
+        description="A timestamp indicating by which time the action must complete. The request may be rejected if the Action Provider does not expect to be able to complete the action before the deadline or if it represents a time greater than the maximum_deadline specified in the Provider Description.",
+    )
+    release_after: Optional[str] = Field(
+        None,
+        description="An ISO8601 time duration value indicating how long retention of the status of the action be retained after it reaches a completed state. Action Providers may limit the maximum value. It is recommended that Providers provide a default release_after value of approximately 30 days.",
+        regex="^P(?!$)(\d+(?:\.\d+)?Y)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?W)?(\d+(?:\.\d+)?D)?(T(?=\d)(\d+(?:\.\d+)?H)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?S)?)?$",
```

Comment:
And nice that the regex can be added here

Reply:
I was impressed that the regex field is actually used to validate incoming strings. So apart from being documentation, it does meaningful validation!